Docs/StyleGuide/Style-Guide.md

```diff
@@ -0,0 +1,280 @@
+```dafny
+import opened StandardLibrary
+...
+```
+
+When there are more than two modules with the same method name, specify the name of the module during method invocation.
+
+```dafny
+import opened MyModule
+import opened YourModule
+...
+method MyMethod() {
+    MyModule.foo();
+    YourModule.foo();
+}
+```
+
+In case typing the full module name is cumbersome, import it with a shorthanded name.
+
+```dafny
+import M = MyModuleWithACumbersomeName
+import Y = YourModuleWithACumbersomeName
+...
+method MyMethod() {
+    M.foo();
+    Y.foo();
+}
+```
+
+Common imports, such as `StandardLibrary` and `Native`, should be grouped together, followed by custom module imports
+with a blank line in-between.
+
+```dafny
+import opened StandardLibrary
+import opened Native
+
+import opened Donut
+import Coffee
+```
+
+## Indentation and Line Breaks
+
+### Tabs or Spaces?
+Spaces are preferred over tabs. Tabs should only be used to remain consistent with the existing code with tabs.
+
+Use 4 spaces for each indentation.
+
+### Maximum Character Limit
+Although there is no strict requirement, it is generally recommended to have maximum of 120 characters in each line.
+
+### Functions, Methods, Predicates, and Lemmas
+Every Dafny method has the following signature.
+```dafny
+method {:<attributes>} MethodName(param1: Type, param2: Type) returns (ret: Type)
+    decreases param1
+    requires P()
+    modifies param2
+    ensures Q()
+```
+
+When possible, put MethodName and `returns` statement in the same line, as the keyword `returns` is distinct from other
+method specification clauses, such as `modifies`, `requires`, `ensures`, and `decreases`, which can be mixed in any
+order (although not recommended for readability), while `returns` should always be right after the parameter
+definitions. Each method specification clause should be in a separate line.
+
+In case the Method signature is too long, we can break it down.
+```dafny
+method {:<attributes>} MethodName(param1: Type, param2: Type,
+        param3: Type, param4: Type, param5: Type)
+    returns (ret1: Type, ret2: Type, ret3: Type, ret4: Type,
+        ret5: Type)
+    decreases param1
+    requires P1()
+    requires P2()
+    requires P3()
+    modifies param2
+    modifies param3
+    ensures Q1()
+    ensures Q2()
+```
+
+Multiple `requires` or `ensures` can be combined into one:
+```dafny
+requires
+    P1() &&
+    P2() &&
+    P3()
+```
+The same rules apply to `function`, `predicate`, and `lemma` definitions.
+
+## Things to Avoid
+
+### Parenthesis
+
+In a lot of cases, Dafny does not require parentheses around expressions. Here are some examples.
+
+* If-Else-While Statements
+```dafny
+// YES
+var i := 1;
+while i < 10 {
+    ...
+    if 1 < i {
+        ...
+    }
+    ...
+}
+
+// NO
+var i := 1;
+while (i < 10) {
+    ...
+    if (1 < i) {
+        ...
+    }
+    ...
+}
+```
+
+* Simple Boolean/Arithmetic Expressions
+```dafny
+// YES
+method Collatz(num: nat)
+    decreases *
+{
```

Comment:
Shouldn't this { be on the previous line?

Reply:
I would say no, because I like the following rule (which already is in this guide):
> In case the method (or function, lemma, etc) signature is too long to fit in one line, or in case the signature has at least one specification clause, the new brace goes on a the new line.

Reply:
I vote for a clear no on this, hard to tell where code starts and the contract ends when not having an empty line there.

Reply:
sg